### PR TITLE
Integrate the MSI and SP and enable cloud agents

### PR DESCRIPTION
**Pull request by LGDoor**

### CODE REVIEW

File: jenkins/install_jenkins.sh [hunk 1]
```diff
@@ -9,8 +10,16 @@ Arguments
...
+  --service_principal_type|-sp        : The type of service principal: MSI or manual.
+  --service_principal_id|-sid         : The service principal ID.
+  --service_principal_secret|-ss      : The service principal secret.
+  --subscription_id|-subid            : The subscription ID of the SP.
+  --tenant_id|-tid                    : The tenant id of the SP.
   --artifacts_location|-al            : Url used to reference other scripts/artifacts.
   --sas_token|-st                     : A sas token needed if the artifacts location is private.
+  --cloud_agents|-ca                  : aci, vm or no
+  --resource_group|-rg                :     
+  --location|-lo                      :
```
Comment (chenkennt) [inline jenkins/install_jenkins.sh]: Add descriptions for the last two arguments?

File: jenkins/install_jenkins.sh [hunk 2]
```diff
@@ -9,8 +10,16 @@ Arguments
...
+  --service_principal_type|-sp        : The type of service principal: MSI or manual.
+  --service_principal_id|-sid         : The service principal ID.
+  --service_principal_secret|-ss      : The service principal secret.
+  --subscription_id|-subid            : The subscription ID of the SP.
+  --tenant_id|-tid                    : The tenant id of the SP.
   --artifacts_location|-al            : Url used to reference other scripts/artifacts.
   --sas_token|-st                     : A sas token needed if the artifacts location is private.
+  --cloud_agents|-ca                  : The type of the cloud agents: aci, vm or no.
+  --resource_group|-rg                : the resource group name.
+  --location|-lo                      : the resource group location.
```
Comment (chenkennt) [inline jenkins/install_jenkins.sh:22]: The resource group and location for cloud agents?

File: jenkins/install_jenkins.sh [hunk 3]
```diff
@@ -240,9 +283,154 @@ inter_jenkins_config=$(sed -zr -e"s|<disabledAdministrativeMonitors/>|{disable-r
...
+  <id>sp</id>
+  <description>Local MSI</description>
+  <msiPort>50342</msiPort>
+</com.microsoft.azure.util.AzureMsiCredentials>
+EOF
+)
+sp_cred=$(cat <<EOF
+<com.microsoft.azure.util.AzureCredentials>
+  <scope>GLOBAL</scope>
+  <id>sp</id>
```
Comment (chenkennt) [inline jenkins/install_jenkins.sh]: make this a more descriptive name

File: jenkins/install_jenkins.sh [hunk 4]
```diff
@@ -240,9 +283,154 @@ inter_jenkins_config=$(sed -zr -e"s|<disabledAdministrativeMonitors/>|{disable-r
...
+    </templates>
+  </com.microsoft.jenkins.containeragents.aci.AciCloud>
+</clouds>
+EOF
+)
+
+agent_admin_cred=$(cat <<EOF
+<com.cloudbees.plugins.credentials.impl.UsernamePasswordCredentialsImpl>
+  <scope>GLOBAL</scope>
+  <id>agent</id>
```
Comment (chenkennt) [inline jenkins/install_jenkins.sh]: same here, use something like agent admin password

File: jenkins/install_jenkins.sh [hunk 5]
```diff
@@ -240,9 +283,154 @@ inter_jenkins_config=$(sed -zr -e"s|<disabledAdministrativeMonitors/>|{disable-r
...
+  <description>the admin account for the vm agents</description>
+  <username>agentadmin</username>
+  <password>1234QWERasdf#</password>
```
Comment (chenkennt) [inline jenkins/install_jenkins.sh]: how can user know the username and password?

File: solution_template/jenkins/createUiDefinition.json [hunk 6]
```diff
@@ -160,6 +136,114 @@
...
+                {
+                  "label": "Manual",
+                  "value": "manual"
+                }
+              ]
+            },
+            "visible": true
+          },
+          {
+            "name": "section1",
```
Comment (chenkennt) [inline solution_template/jenkins/createUiDefinition.json]: better name

File: jenkins/install_jenkins.sh (hunk 2 above)
Comment (LGDoor) [reply jenkins/install_jenkins.sh:22]: for jenkins master, while the agents are configured in the same resource group by default.

**Review by chenkennt — APPROVED** [review]

**Pull request merged**